pkg/channel/sink.go

```diff
@@ -0,0 +1,85 @@
+package channel
+
+import (
+	"errors"
+	"sync/atomic"
+)
+
+var ErrSinkSealed = errors.New("The channel is sealed")
+
+// SinkWithSender is a helper struct that allows to send messages to a message sink.
+// The SinkWithSender abstracts the message sink which has a certain sender, so that
+// the sender does not have to be specified every time a message is sent.
+// At the same it guarantees that the caller can't alter the `sender`, which means that
+// the sender can't impersonate another sender (and we guarantee this on a compile-time).
+type SinkWithSender[SenderType comparable, MessageType any] struct {
+	// The sender of the messages. This is useful for multiple-producer-single-consumer scenarios.
+	sender SenderType
+	// The message sink to which the messages are sent.
+	messageSink chan<- Message[SenderType, MessageType]
+	// A channel that is used to indicate that our channel is considered sealed. It's akin
+	// to a close indication without really closing the channel. We don't want to close
+	// the channel here since we know that the sink is shared between multiple producers,
+	// so we only disallow sending to the sink at this point.
+	sealed chan struct{}
+	// A "mutex" that is used to protect the act of closing `sealed`.
+	alreadySealed atomic.Bool
+}
+
+// Creates a new MessageSink. The function is generic allowing us to use it for multiple use cases.
+// Note that since the current implementation accepts a channel, it's **not responsible** for closing it.
+func NewSink[S comparable, M any](sender S, messageSink chan<- Message[S, M]) *SinkWithSender[S, M] {
+	return &SinkWithSender[S, M]{
+		sender:      sender,
+		messageSink: messageSink,
+		sealed:      make(chan struct{}),
+	}
+}
+
+// Sends a message to the message sink. Blocks if the sink is full!
+func (s *SinkWithSender[S, M]) Send(message M) error {
+	if s.alreadySealed.Load() {
+		return ErrSinkSealed
+	}
+
+	messageWithSender := Message[S, M]{
+		Sender:  s.sender,
+		Content: message,
+	}
+
+	select {
+	case <-s.sealed:
+		return ErrSinkSealed
+	case s.messageSink <- messageWithSender:
+		return nil
+	}
```

Comment:
You do not know which of the two cases is executed at runtime. With 'default' you make sure that the case is checked first.
```go
select {
	case <-s.sealed:
		return ErrSinkSealed
	default: 
	s.messageSink <- messageWithSender:
		return nil
	}
```

Reply:
But is it not different semantically? I.e. what if the sender gets to the `s.messageSink <- messageWithSender` part and blocks there waiting for the reader to get ready? We must ensure that once the reader is not ready to accept new messages (e.g. not interested in them anymore), then we don't continue waiting on the channel. Essentially that's what I tried to describe, i.e. it's semantically I wanted this

```go
select {
    case <-done:
        return OkRecipientDoesNotExpectNewMessages
    case s.messageSink <- messageWithSender:
        return nil
}
```

Though your concern is valid, I tried to catch this case by checking the atomic variable at the beginning of the function (this does not guarantee that those who called `Send()` just before `Close()` won't send a value though).

Reply:
Ok got it better send a message more than blocking.